idaes/surrogate/plotting/sm_plotter.py

```diff
@@ -60,64 +68,59 @@ def scatter2D(xdata, zdata, xtest, zfit, xlabels=None, zlabels=None, dz=None,
         zlabels = []
         for j in range(numouts):
             zlabels.append('z' + str(j+1))
-    if clabel is None:
-        clabel = 'Confidence Intervals'
 
-    pointidx = -1  # dummy index for each plot produced
-    fig, ax = [], []
+    if filename is not None:
+        if filename[-4:] != '.pdf':  # adding extension if not present
+            name = filename + '.pdf'
+        else:
+            name = filename
+        pdf = PdfPages(name)
+
     for j in range(numouts):  # loop over all outputs, zj
         for i in range(numins):  # plot every possible zj = f(xi)
-            pointidx += 1
-            fig.append(plt.figure())
-            ax.append(fig[pointidx].add_subplot())
-            ax[pointidx].scatter(xdata[:, i], zdata[:, j], c='grey', marker='s',
-                                 label='Data')
-            ax[pointidx].scatter(xtest[:, i], zfit[:, j], c='b',
-                                 marker='.', label='Model')
-
-            # plot confidence intervals if given data
-            if clo is not None and chi is not None:
-                ax[pointidx].scatter(xdata[:, i], clo[:, j], c='g',
-                                     marker='.', label=clabel)
-                ax[pointidx].scatter(xdata[:, i], chi[:, j], c='g',
-                                     marker='.', label=clabel)
-
-            ax[pointidx].set_xlabel(xlabels[i])
-            ax[pointidx].set_ylabel(zlabels[j])
-            ax[pointidx].set_title('2D Scatter Plot')
-            ax[pointidx].legend()
-
-            if dz is not None and dzfit is not None:
-                pointidx += 1
-                fig.append(plt.figure())
-                ax.append(fig[pointidx].add_subplot())
-                ax[pointidx].scatter(xdata[:, i], dz[:, j], c='grey', marker='s',
-                                     label='Data')
-                ax[pointidx].scatter(xdata[:, i], dzfit[:, j], c='b',
-                                     marker='.', label='Model')
-                ax[pointidx].set_xlabel(xlabels[i])
-                ax[pointidx].set_ylabel('d[' + str(zlabels[j]) + ']/d[' +
-                                        xlabels[i] + ']')
-                ax[pointidx].set_title('2D Derivative Plot')
-                ax[pointidx].legend()
+            fig = plt.figure()
```

Comment:
I am not completely familiar with the plotting tools, but this looks to me that this keeps all the plots open which will result in the memory issues @carldlaird was discussing.

Reply:
The new form of the loop keeps the file open, but overwrites the figure/axis on each iteration rather than saving a list of figure objects as before. If we close the file after saving each figure, PdfPages() will overwrite the previous file and plot. There is another package (PyPDF) that can open files and add new pages, but it would require an additional dependency since it is not part of any already installed package.

Reply:
This is exactly what I wanted. Thanks, @bpaul4